### PR TITLE
Have tests be built into libs

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

Reduces link time of the `RunTests` executable by having the tests be compiled into libraries that are then linked into the executable.  On my machine this took link time of `RunTests` from ~6.1 seconds to ~1.9 seconds.

What breaks:
- Any new test source files (`cpp`) will need to be added to the new lists.

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [x] Build system
- [ ] Continuous integration

### Code review checklist

- Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- Code has documentation and unit tests
- Private member variables have a trailing underscore
- Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- Header order:
  1. hpp corresponding to cpp (only in cpp files)
  2. Blank line (only in cpp files)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- File lists in CMake are alphabetical
- Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- Mark objects `const` whenever possible
- Almost always `auto`, except with expression templates, i.e. `DataVector`
- All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."
- Prefix commits addressing PR requests with `fixup`. These commits are flagged
  by Travis so we do not accidentally merge them.
- Include what you use, prefer forward declarations in hpp files.
- Explicitly make numbers floating point, e.g. `2.` or `2.0` over `2`
- Use `Tensor`'s non-member get if the indices are constant expressions, e.g.
  `get<1>(tensor)`

### CODE REVIEW

File: cmake/SpectreAddCatchTests.cmake [hunk 1]
```diff
@@ -40,12 +40,23 @@
...
+      set(CORRECTED_LIB_SOURCES "")
+      get_target_property(LIB_SOURCES ${TEST_LIB} SOURCES)
+      get_target_property(LIB_FOLDER ${TEST_LIB} FOLDER)
+      foreach (SOURCE ${LIB_SOURCES})
+        set(CORRECTED_LIB_SOURCES "${CORRECTED_LIB_SOURCES};${LIB_FOLDER}${SOURCE}")
+      endforeach()
+      set(SOURCE_FILES "${SOURCE_FILES};${CORRECTED_LIB_SOURCES}")
+    endforeach()
+
+
```
Comment (nilsdeppe) [inline cmake/SpectreAddCatchTests.cmake]: remove extra blank line

File: cmake/SpectreAddCatchTests.cmake [hunk 2]
```diff
@@ -40,12 +40,23 @@
...
 
 # Main function - the only one designed to be called from outside this module.
-function(spectre_add_catch_tests TEST_TARGET)
+function(spectre_add_catch_tests TEST_TARGET TEST_LIBS)
     get_target_property(SOURCE_FILES ${TEST_TARGET} SOURCES)
     # For each of the source files, we use spectre_parse_file to find all the
     # Catch tests inside the source file and add them to CTest.
     # We ignore the Charm++ generated header files.
 
+    foreach (TEST_LIB ${TEST_LIBS})
```
Comment (nilsdeppe) [inline cmake/SpectreAddCatchTests.cmake]: add comment

File: tests/Unit/TestingFramework.hpp [hunk 3]
```diff
@@ -20,6 +20,15 @@
 #include "Utilities/Requires.hpp"
 #include "Utilities/TypeTraits.hpp"
 
+/// \cond
+// The macro SPECTRE_TEST_REGISTER_FUNCTION is defined inside the add_test
+// library CMake function. It is used to make a call into a translation unit so
```
Comment (nilsdeppe) [inline tests/Unit/TestingFramework.hpp]: add `_` between `add_test` and `library`

File: tests/Unit/CMakeLists.txt [hunk 4]
```diff
@@ -7,8 +7,6 @@ set(SPECTRE_TESTS "TestFramework.cpp")
 
 set(SPECTRE_COMPILATION_TESTS "TestCompilationFramework.cpp")
 
-set(LIBRARIES_TO_LINK "Informer")
-
```
Comment (nilsdeppe) [inline tests/Unit/CMakeLists.txt]: follow instructions and zero the global properties

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/WaveEquation/CMakeLists.txt [hunk 5]
```diff
@@ -1,11 +1,19 @@
 # Distributed under the MIT License.
 # See LICENSE.txt for details.
 
-set(SPECTRE_UNIT_WAVE_EQUATION
-  PointwiseFunctions/AnalyticSolutions/WaveEquation/Test_PlaneWave.cpp
+
+set(LIBRARY "Test_WaveEquation")
+
+set(TEST_DIR
```
Comment (nilsdeppe) [inline tests/Unit/PointwiseFunctions/AnalyticSolutions/WaveEquation/CMakeLists.txt]: inline TEST_DIR do so liberally

File: tests/Unit/RunTests.cpp [hunk 6]
```diff
@@ -12,8 +12,10 @@
 #include "Parallel/Abort.hpp"
 #include "Parallel/Exit.hpp"
 #include "Parallel/Printf.hpp"
+#include "tests/Unit/RunTestsRegister.hpp"
 
 RunTests::RunTests(CkArgMsg* msg) {
+  register_run_tests_libs();
```
Comment (nilsdeppe) [inline tests/Unit/RunTests.cpp]: move after terminate thing

**Review by kidder — DISMISSED** [review]

**Comment by nilsdeppe** [comment]
This has been rebased on develop, updated to handle the new tests correctly, and also addressed the review comments.

The changes for the new tests are:
- Pypp has its own test lib `Test_Pypp`
- Newtonian Euler has its own test lib `Test_NewtonianEuler`

**Review by kidder — APPROVED** [review]

**Pull request merged**